File: lib/features/payments/data/payment_repository.dart
import 'package:flutter_riverpod/flutter_riverpod.dart';

import '../../../core/mock/mock_network.dart';
import 'payment.dart';

abstract interface class PaymentRepository {
  Future<void> pay(Payment payment);
}

class FakePaymentRepository implements PaymentRepository {
  FakePaymentRepository(this._network);

  final MockNetwork _network;

  @override
  Future<void> pay(Payment payment) async {
    await _network.delay();
    _network.countAndMaybeFail();
  }
}

final paymentRepositoryProvider = Provider<PaymentRepository>(
  (ref) => FakePaymentRepository(ref.watch(mockNetworkProvider)),
);

File: src/features/payments/data/paymentRepository.ts
import type { MockNetwork } from '../../../core/mock/mockNetwork';
import type { Payment } from './payment';

export interface PaymentRepository {
  pay(payment: Payment): Promise<void>;
}

export class FakePaymentRepository implements PaymentRepository {
  constructor(private readonly network: MockNetwork) {}

  async pay(_payment: Payment): Promise<void> {
    await this.network.delay();
    this.network.countAndMaybeFail();
  }
}
